test/integration/framework/etcd.go

```diff
@@ -182,37 +182,47 @@ func EtcdMain(tests func() int) {
 	// Bail out early when -help was given as parameter.
 	flag.Parse()
 
-	before := runtime.NumGoroutine()
+	// Must be called *before* creating new goroutines.
+	goleakOpts := IgnoreBackgroundGoroutines()
+
+	goleakOpts = append(goleakOpts,
+		// lumberjack leaks a goroutine:
+		// https://github.com/natefinch/lumberjack/issues/56 This affects tests
+		// using --audit-log-path (like
+		// ./test/integration/apiserver/admissionwebhook/reinvocation_test.go).
+		// In normal production that should be harmless. We don't know here
+		// whether the test is using that, so we have to suppress reporting
+		// this leak for all tests.
+		//
+		// Both names occurred in practice.
+		goleak.IgnoreTopFunction("k8s.io/kubernetes/vendor/gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun"),
+		goleak.IgnoreTopFunction("gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun"),
+	)
+
 	stop, err := startEtcd()
 	if err != nil {
 		klog.Fatalf("cannot run integration tests: unable to start etcd: %v", err)
 	}
 	result := tests()
 	stop() // Don't defer this. See os.Exit documentation.
 
-	checkNumberOfGoroutines := func() (bool, error) {
-		// We leave some room for leaked goroutines as there are
-		// still some leaks, mostly:
-		// - leak from lumberjack package we're vendoring
-		// - leak from apiserve healthz
-		// - leak from opencensus library
-		// Once fixed, we should be able to bring it down to zero.
-		if dg := runtime.NumGoroutine() - before; dg <= 3 {
-			return true, nil
+	// Several tests don't wait for goroutines to stop. goleak.Find retries
+	// internally, but not long enough. 5 seconds seemed to be enough for
+	// most tests, even when testing in the CI.
+	timeout := 5 * time.Second
+	start := time.Now()
+	for {
+		err := goleak.Find(goleakOpts...)
```

Comment:
This loop can get removed if https://github.com/uber-go/goleak/pull/94 gets merged.

Reply:
exactly 👍

Reply:
looks like that PR got merged, can we drop this loop now?

Reply:
oh, looks like that didn't make it into the 1.2.1 tag

Reply:
I'll keep an eye on that and update once it's in a release.